### PR TITLE
Update to use apischema

### PR DESCRIPTION
**Pull request by thomascobb**

### CODE REVIEW

**Comment by thomascobb** [comment]
Python3,.9 failing at the moment, issue https://github.com/wyfo/apischema/issues/279 reported

File: src/pvi/_produce/base.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    R = "R"  #: Read record only
+    W = "W"  #: Write record only
+    RW = "RW"  #: Read and write record
+
+    def needs_read_record(self):
+        return self != self.W
+
+    def needs_write_record(self):
+        return self != self.R
```
Comment (GDYendell) [inline src/pvi/_produce/base.py:30]: These should be `@property`s

File: src/pvi/_schema_utils.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+
+# Permanently cache so we don't include deserialization subclasses defined below
+@lru_cache(maxsize=None)
+def rec_subclasses(cls: Cls) -> List[Cls]:
+    """Recursive implementation of type.__subclasses__"""
+    subclasses = []
+    for sub_cls in cls.__subclasses__():
+        subclasses += [sub_cls] + rec_subclasses(sub_cls)
+    return subclasses
```
Comment (GDYendell) [inline src/pvi/_schema_utils.py:22]: Could we have some notes about what this is for?

**Review by GDYendell — COMMENTED** [review]
I have just commented on everything, I but a aware some of it is not things you have changed recently.

File: src/pvi/__main__.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    )
+    output.write_text(json.dumps(schema, indent=2))
+
+
+@app.command()
+def produce(
+    output: Path = typer.Argument(..., help="filename to write the product to"),
+    producer: Path = typer.Argument(..., help="path to the producer .pvi.yaml file"),
+):
+    """Create template/csv/device product from producer YAML"""
```
[the remark this answers is not in the document]
Comment (thomascobb) [reply src/pvi/__main__.py:70]: I've added other to this

File: src/pvi/_convert/_asyn_convert.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            ]
+        try:
+            idx = type_fields.index((self.type, self.fields["DTYP"]))
+            asyn_class = asyn_components[idx]
+            return asyn_class
+        except ValueError as e:
+            raise ValueError(
+                f"{self.name} asyn type: ({self.type}, {self.fields['DTYP']}) "
+                f"not found in {type_fields}"
+            ) from e
```
[the remark this answers is not in the document]
Comment (thomascobb) [reply src/pvi/_convert/_asyn_convert.py:75]: includes the traceback of e

File: src/pvi/_convert/_asyn_convert.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+    def are_clashes(self) -> bool:
+        empty = not self._get_field_clashes()
+        return not empty
+
+    def generate_component(self) -> AsynParameter:
+        asyn_class = self.write_record.asyn_component_type()
+
+        non_default_args: Dict[str, Any] = dict()
+        non_default_args["demand_auto_updates"] = self._get_demand_auto_updates(
```
[the remark this answers is not in the document]
Comment (thomascobb) [reply src/pvi/_convert/_asyn_convert.py:152]: maybe, but I'll leave that to someone else to look at...

File: src/pvi/_convert/_template_convert.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+class ParameterRoleMatcher:
+    @staticmethod
+    def get_actions(
+        read_records: List[AsynRecord], write_records: List[AsynRecord]
+    ) -> List[Action]:
+        actions = [
+            Action(write_record=w)
+            for w in write_records
+            if w.get_parameter_name()
+            not in [r.get_parameter_name() for r in read_records]
```
[the remark this answers is not in the document]
Comment (thomascobb) [reply src/pvi/_convert/_template_convert.py:282]: not sure what you mean...

File: src/pvi/_format/utils.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+
+
+@dataclass
+class Bounds:
+    x: int = 0
+    y: int = 0
+    w: int = 0
+    h: int = 0
+
+    def copy(self) -> "Bounds":
```
[the remark this answers is not in the document]
Comment (thomascobb) [reply src/pvi/_format/utils.py:36]: It's a forward reference to the class Bounds, which doesn't exist yet as it's a method of that class

File: src/pvi/_format/utils.py [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+            def format(self) -> List[T]:
+                padding = sized(self.bounds)
+                texts: List[T] = []
+                if search == GroupType.SCREEN:
+                    properties = {k: getattr(self, v) for k, v in attrs.items()}
+                    texts.append(
+                        template.set(template.screen, self.bounds, **properties)
+                    )
+                # TODO: group things?
+                if make_widgets:
```
[the remark this answers is not in the document]
Comment (thomascobb) [reply src/pvi/_format/utils.py:152]: All callables are truthy so makes no difference here

File: src/pvi/_produce/base.py (hunk 1 above)
Comment (thomascobb) [reply src/pvi/_produce/base.py:30]: could be, but I tend not to put in properties unless I'm trying to maintain backwards compatibility

File: src/pvi/_schema_utils.py (hunk 2 above)
Comment (thomascobb) [reply src/pvi/_schema_utils.py:22]: rec_subclasses() or the whole of this file?

File: src/pvi/_format/utils.py [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+class Bounds:
+    x: int = 0
+    y: int = 0
+    w: int = 0
+    h: int = 0
+
+    def copy(self) -> "Bounds":
+        return Bounds(self.x, self.y, self.w, self.h)
+
+    def split(self, w: int, spacing: int) -> Tuple["Bounds", "Bounds"]:
```
[the remark this answers is not in the document]
Comment (thomascobb) [reply src/pvi/_format/utils.py]: changed

File: src/pvi/_convert/_parameters.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    def _get_read_record_suffix(self) -> Optional[str]:
+        raise NotImplementedError(self)
+
+
+class WriteParameterMixin:
+    def _get_initial(self, write_record: Record) -> str:
+        try:
+            pini = write_record.fields["PINI"].lower()
+        except KeyError:
+            pini = "no"
```
[the remark this answers is not in the document]
Comment (thomascobb) [reply src/pvi/_convert/_parameters.py]: changed

File: src/pvi/_schema_utils.py (hunk 2 above)
Comment (GDYendell) [reply src/pvi/_schema_utils.py:22]: Just `rec_subclasses()`

**Pull request merged**